sdk/lib/udmihelp/udmihelp.c

```diff
@@ -0,0 +1,162 @@
+
+/* Load SMBIOS Data */
+PVOID
+LoadSMBiosData(
+    _Inout_updates_(ID_STRINGS_MAX) PCHAR * Strings)
+{
+    HANDLE hKernel;
+    PVOID pBuffer = NULL;
+    HKEY hKey;
+    DWORD dwType, dwCheck, dwBytes = 0;
+
+    hKernel = GetModuleHandleW(L"kernel32.dll");
+    if (!hKernel)
+        return NULL;
+
+    /* Try using GetSystemFirmwareTable (works on NT 5.2 and higher) */
+    pGetSystemFirmwareTable = (void *)GetProcAddress(hKernel, "GetSystemFirmwareTable");
+    if (pGetSystemFirmwareTable)
+    {
+        dwBytes = pGetSystemFirmwareTable('RSMB', 0, NULL, 0);
+        if (dwBytes > 0)
+        {
+            pBuffer = malloc(dwBytes);
+            if (!pBuffer)
+            {
+                return NULL;
+            }
+            dwCheck = pGetSystemFirmwareTable('RSMB', 0, pBuffer, dwBytes);
+            if (dwCheck != dwBytes)
+            {
+                free(pBuffer);
+                return NULL;
+            }
+        }
+    }
+    if (dwBytes == 0)
+    {
+        /* Try using registry (works on NT 5.1) */
+        if (RegOpenKeyW(HKEY_LOCAL_MACHINE,
+                        L"SYSTEM\\CurrentControlSet\\Services\\mssmbios\\Data",
+                        &hKey) != ERROR_SUCCESS)
+        {
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             NULL,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        pBuffer = malloc(dwBytes);
+        if (!pBuffer)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             pBuffer,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            free(pBuffer);
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        RegCloseKey(hKey);
+    }
+    ParseSMBiosTables(pBuffer, dwBytes, Strings);
+    return pBuffer;
+}
+
+/* Trim converted DMI string */
+VOID
+TrimDmiStringW(
+    _Inout_ LPWSTR lpStr)
+{
+    SIZE_T Length;
+    INT i = 0;
+
+    if (!lpStr)
+        return;
+
+    Length = wcslen(lpStr);
+    if (Length == 0)
+        return;
+
+    /* Trim leading spaces */
+    while (i < Length && lpStr[i] <= L' ')
+    {
+        i++;
+    }
+
+    if (i > 0)
+    {
+        Length -= i;
+        memmove(lpStr, lpStr + i, (Length + 1) * sizeof(WCHAR));
+    }
+
+    /* Trim trailing spaces */
+    while (Length && lpStr[Length-1] <= L' ')
+    {
+        lpStr[Length-1] = L'\0';
+        --Length;
+    }
+}
+
+/* Convert string from SMBIOS */
+SIZE_T
+GetSMBiosStringW(
+    _In_ PCHAR DmiString,
```

Comment:
Is this always a NULL-terminated string? If so, please use PWSTR (and if also it's a constant string, use PCWSTR).

Reply:
Yes, but it's not wide string. Replaced with `PCSTR`.